typing/ctype.ml

```diff
@@ -238,31 +238,81 @@ let none = newty (Ttuple [])                (* Clearly ill-formed type *)
 
 (**** unification mode ****)
 
-type unification_mode =
-  | Expression (* unification in expression *)
-  | Pattern (* unification in pattern which may add local constraints *)
-
 type equations_generation =
   | Forbidden
   | Allowed of { equated_types : TypePairs.t }
 
+type unification_mode =
+  | Expression (* unification in expression *)
+  | Pattern of
+      { equations_generation : equations_generation;
+        assume_injective : bool;
+        allow_recursive_equations : bool; }
```

Comment:
Personally I find `allow_recursive_equations` clearer than `in_counter_example`.

> If rectypes is true, then recursive equations are allowed even if allow_recursive_equations is false.

I think this is clear from the implementation of the getter function, that uses `!Clflags.recursive_types || ...`.

Reply:
This does not change the fact that this is strange that setting `allow_recursive_equations` to false does not prevent them. I.e., there is a mismatch between the name and the effect.
I suppose that your point is that it is not nice to bring concepts from `Typecore` into `Ctype`.
However, in that case I found it hard to track when this flag was set, and it only makes sense in counter-examples.
I tried to come up with a better name, but couldn't.

Here the property we are enforcing is that abstract types should be unifiable with anything. It would still be correct to just add nothing when the equation is invalid (i.e. recursive without `!Clflags.recursive_types`), but adding an equation, even an invalid one, may allow to reject more impossible cases, so we choose rather to do that. It is OK because the inferred type will not be used anyway.